### PR TITLE
Add support for Objective-C protocols

### PR DESCRIPTION
**Pull request by jacob-carlborg**
This is quite a large change. I've already made 10+ pull requests to prepare for this. I don't think it make sense to break it down into smaller PRs. I've made several commits, I recommend looking at the individual commits for easier review.

Depends on https://github.com/dlang/druntime/pull/3221.
Spec PR: https://github.com/dlang/dlang.org/pull/2864.

Todo:
- [x] Add a test for an interface with a static method with a body
- [x] Mention something about class/static methods in protocols in the changelog
- [x] Update headers
- [x] Create a spec PR

### CODE REVIEW

File: src/dmd/backend/machobj.d [hunk 1]
```diff
@@ -1785,6 +1785,12 @@ int Obj_comdat(Symbol *s)
         flags = S_COALESCED | S_ATTR_PURE_INSTRUCTIONS | S_ATTR_SOME_INSTRUCTIONS;
         s.Sseg = getsegment2(seg_textcoal_nt, sectname, segname, align_, flags);
     }
+    else if ((s.ty() & mTYLINK) == mTYweakLinkage)
```
Comment (jacob-carlborg) [inline src/dmd/backend/machobj.d:1788]: I'm not sure if this is the best solution. What I need is a symbol with weak linkage, stored in a specific section. DMD seems to implement weak linkage using the `SCcomdat ` storage class: https://github.com/dlang/dmd/blob/5fda4ea088f70afb99818482e8482771758136a2/src/dmd/backend/cdef.d#L912
The problem with that is that it will hardcode a section for the symbol.

File: src/dmd/dsymbolsem.d [hunk 2]
```diff
@@ -4870,47 +4875,6 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
 
     override void visit(ClassDeclaration cldec)
     {
-        /// Checks that the given class implements all methods of its interfaces.
-        static void checkInterfaceImplementations(ClassDeclaration cd)
```
Comment (jacob-carlborg) [inline src/dmd/dsymbolsem.d:4874]: This code is just moved the to the second semantic phase.

File: src/dmd/objc.d [hunk 3]
```diff
@@ -499,22 +531,9 @@ extern(C++) private final class Supported : Objc
...
-        // don't report deprecations for the metaclass to avoid duplicated
-        // messages.
-        if (id.objc.isMeta)
-            return;
-
-        id.deprecation("Objective-C interfaces have been deprecated");
-        deprecationSupplemental(id.loc, "Representing an Objective-C class " ~
-            "as a D interface has been deprecated. Please use "~
-            "`extern (Objective-C) extern class` instead");
+        return cd.parent.toPrettyChars(qualifyTypes);
```
Comment (jacob-carlborg) [inline src/dmd/objc.d:536]: I'm not sure if this is the best way. Before this change, if a static method in an interface is not implemented in the class, the error message will include the name of the metaclass (something like `__anonclass_NSObject_Meta2`). This is an implementation detail which I don't want to expose. With this change, it will make sure to print the name of the class and not the metaclass.

File: src/dmd/root/array.d [hunk 4]
```diff
@@ -1013,7 +1038,7 @@ enum isInputRange(R) =
 
 /// Evaluates to `true` if `func` can be called with a value of `T` and returns `bool`.
 enum isPredicateOf(alias func, T) =
-    is(ReturnType!(typeof((T t) => func(t))) == bool);
+    __traits(compiles, { auto _ = (T t) { if (func(t)) {} }; });
```
Comment (jacob-carlborg) [inline src/dmd/root/array.d:1041]: This change makes sure everything that is convertible to `bool` will be accepted, instead of just `bool`.

File: src/dmd/semantic2.d [hunk 5]
```diff
@@ -602,6 +602,66 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
+        if (cd.semanticRun >= PASS.semantic2done)
+            return;
+        assert(cd.semanticRun <= PASS.semantic2);
+        cd.semanticRun = PASS.semantic2;
+
+        checkInterfaceImplementations(cd);
+        visit(cast(AggregateDeclaration) cd);
+    }
+
+    override void visit(InterfaceDeclaration cd)
```
Comment (jacob-carlborg) [inline src/dmd/semantic2.d:661]: This will make sure the above code is not run for interfaces, only for classes.

File: test/fail_compilation/diag8318.d [hunk 6]
```diff
@@ -1,7 +1,12 @@
 /*
 TEST_OUTPUT:
 ---
-fail_compilation/diag8318.d(13): Error: function `diag8318.Bar8318.foo` return type inference is not supported if may override base class function
+fail_compilation/diag8318.d(18): Error: function `diag8318.Bar8318.foo` return type inference is not supported if may override base class function
+fail_compilation/diag8318.d(23): Error: function `diag8318.C10021.makeI` return type inference is not supported if may override base class function
+fail_compilation/diag8318.d(31): Error: function `diag8318.Bar10195.baz` return type inference is not supported if may override base class function
+fail_compilation/diag8318.d(37): Error: function `diag8318.B14173.foo` does not override any function
+fail_compilation/diag8318.d(23): Error: class `diag8318.C10021` interface function `I10021 makeI()` is not implemented
+fail_compilation/diag8318.d(29): Error: class `diag8318.Bar10195` interface function `int baz()` is not implemented
```
Comment (jacob-carlborg) [inline test/fail_compilation/diag8318.d:9]: This change was necessary because the order of these error messages changed, when the code which checks if an interface is implemented or not was moved to the second phase of semantic.

File: src/dmd/root/array.d (hunk 4 above)
Comment (MoonlightSentinel) [inline src/dmd/root/array.d:1041]:
```suggestion
    is(typeof((T t) => func(t) ? 0 : 0)));
```

Comment (BorisCarvajal) [reply src/dmd/root/array.d:1041]: or maybe:
```suggestion
    is(typeof(!func(T.init)));
```

Comment (jacob-carlborg) [reply src/dmd/root/array.d:1041]: @BorisCarvajal I did that before. Apparently it doesn't support `ref` delegates: https://github.com/dlang/dmd/pull/11532#discussion_r467517734.

Comment (BorisCarvajal) [reply src/dmd/root/array.d:1041]: I just tested that works.
https://run.dlang.io/is/skNYh7

Comment (MoonlightSentinel) [reply src/dmd/root/array.d:1041]: Neither of those delegates accept the parameter by ref ...

Comment (BorisCarvajal) [reply src/dmd/root/array.d:1041]: >parameter by ref

Thanks for pointing out that.

Comment (jacob-carlborg) [reply src/dmd/root/array.d]: Fixed.

**Comment by jacob-carlborg** [comment]
No one dares to have a look 😃 ?

Do we need Walter to have a look at the backend changes?

**Comment by 12345swordy** [comment]
@WalterBright please review this as this pr have modify the back end.

**Comment by 12345swordy** [comment]
Some test suits are failing for some reason.

**Comment by jacob-carlborg** [comment]
> Some test suits are failing for some reason.

* macOS is obviously failing since the druntime PR has not been merged
* SemaphoreCI (GDC) is probably failing due to an ancient D host compiler (I'll try to fix that)
* Azure pipelines (Windows) is failing because it was trying to run some Objective-C tests. I've fixed that now

**Comment by jacob-carlborg** [comment]
@12345swordy tests (except macOS, which is expected) are passing now.

**Comment by 12345swordy** [comment]
@jacob-carlborg that is great. Now we need to ping @WalterBright to review the changes that you made regarding the backend.

**Comment by 12345swordy** [comment]
Ping!

**Comment by jacob-carlborg** [comment]
All tests are passing now since the druntime PR has been merged.

**Comment by 12345swordy** [comment]
@thewilsonator can this be merge now?

File: changelog/objc_protocols.dd [hunk 7]
```diff
@@ -0,0 +1,73 @@
+Add support for Objetive-C protocols
```
Comment (WalterBright) [inline changelog/objc_protocols.dd:1]: spelling

File: changelog/objc_protocols.dd [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+It's now possible to declare Objective-C protocols. In D, this is declared using
```
Comment (WalterBright) [inline changelog/objc_protocols.dd:3]: A link to what a protocol is would be nice, I have no idea what it is.

File: src/dmd/backend/ty.d [hunk 9]
```diff
@@ -166,6 +166,7 @@ enum
 
     // Used for symbols going in the __thread_data section for TLS variables for Mach-O 64bit
     mTYthreadData   = 0x5000,
+    mTYweakLinkage  = 0x6000,
```
Comment (WalterBright) [inline src/dmd/backend/ty.d:169]: I don't think mTYweakLinkage is used for symbols going in the __thread_data section? Some description of it should be here.

File: src/dmd/hdrgen.d [hunk 10]
```diff
@@ -3411,6 +3411,8 @@ private void visitFuncIdentWithPostfix(TypeFunction t, const char[] ident, OutBu
         linkageToBuffer(buf, t.linkage);
         buf.writeByte(' ');
     }
+    if (t.linkage == LINK.objc && isStatic)
+        buf.write("static ");
```
Comment (WalterBright) [inline src/dmd/hdrgen.d:3415]: why do other static member functions not need this?

File: src/dmd/toobj.d [hunk 11]
```diff
@@ -449,6 +449,12 @@ void toObjFile(Dsymbol ds, bool multiobj)
             // Put out the members
             id.members.foreachDsymbol( (s) { visitNoMultiObj(s); } );
 
+            // Objetive-C protocols are only outputted if they're implemented by
```
Comment (WalterBright) [inline src/dmd/toobj.d:452]:
```
Objective-C protocols are only output if implemented
```

File: src/dmd/toobj.d [hunk 12]
```diff
@@ -449,6 +449,12 @@ void toObjFile(Dsymbol ds, bool multiobj)
...
+            // a class. If they are implemented by a class, they're outputted
```
Comment (WalterBright) [inline src/dmd/toobj.d:453]:
```
as a class. If so, they're output
```

File: src/dmd/toobj.d [hunk 13]
```diff
@@ -449,6 +449,12 @@ void toObjFile(Dsymbol ds, bool multiobj)
...
+            // via the class declaration.
```
Comment (WalterBright) [inline src/dmd/toobj.d:454]: This seems strange. What does "via the class declaration" mean?

File: src/dmd/hdrgen.d (hunk 10 above)
Comment (jacob-carlborg) [reply src/dmd/hdrgen.d:3415]: I have not been able to come up with a case were this would occur for non-objc methods.

File: changelog/objc_protocols.dd (hunk 8 above)
Comment (jacob-carlborg) [reply changelog/objc_protocols.dd:3]: I'll add a link. Short answer: like an interface in D.

File: src/dmd/toobj.d (hunk 13 above)
Comment (jacob-carlborg) [reply src/dmd/toobj.d:454]: The class symbol contains a reference to the protocol symbol. When the class symbol is output, it will automatically output the protocol symbol as well. If the protocol is never implemented, it won't be output.

File: changelog/objc_protocols.dd (hunk 7 above)
Comment (jacob-carlborg) [reply changelog/objc_protocols.dd]: Fixed.

File: changelog/objc_protocols.dd (hunk 8 above)
Comment (jacob-carlborg) [reply changelog/objc_protocols.dd]: Fixed.

File: src/dmd/toobj.d (hunk 11 above)
Comment (jacob-carlborg) [reply src/dmd/toobj.d]: Fixed.

File: src/dmd/toobj.d (hunk 12 above)
Comment (jacob-carlborg) [reply src/dmd/toobj.d]: Fixed.

File: src/dmd/backend/ty.d (hunk 9 above)
Comment (jacob-carlborg) [reply src/dmd/backend/ty.d:173]: Fixed.

**Comment by 12345swordy** [comment]
IS this good to go now?

**Comment by jacob-carlborg** [comment]
I feel that I have addressed all of Walter's concerns.

**Comment by 12345swordy** [comment]
I can't mark this auto-merge, sadly.

File: src/dmd/objc.d [hunk 14]
```diff
@@ -161,12 +163,11 @@ extern (C++) struct ObjcClassDeclaration
     ClassDeclaration metaclass;
 
     /// List of non-inherited methods.
-    FuncDeclarations* methodList;
+    FuncDeclaration[] methodList;
```
Comment (MoonlightSentinel) [inline src/dmd/objc.d:166]: Unrelated change

Comment (jacob-carlborg) [reply src/dmd/objc.d:166]: I don't think it's unrelated. I changed the type because it's easier to operate on a D array than `Array`.

Comment (MoonlightSentinel) [reply src/dmd/objc.d:166]: Fair enough, allthough `Array` is used by default in dmd (and it supports `opSlice`).

File: src/dmd/objc.d [hunk 15]
```diff
@@ -570,6 +589,9 @@ extern(C++) private final class Supported : Objc
     }
     do
     {
+        if (fd.toParent.isInterfaceDeclaration && fd.isFinal)
+            return false;
```
Comment (MoonlightSentinel) [inline src/dmd/objc.d:593]: No test for this case AFAICT.

**Review by MoonlightSentinel — COMMENTED** [review]
Frontend changes & test coverage seem fine (but I can't comment on the glue/backend changes)

File: src/dmd/objc.d (hunk 14 above)
Comment (jacob-carlborg) [reply src/dmd/objc.d:166]: Why do you think that is? Could it possibly have something to do with the fact that DMD used to be written in C++ 😉.

File: src/dmd/objc.d (hunk 15 above)
Comment (jacob-carlborg) [reply src/dmd/objc.d:593]: It's covered by this existing test [1]. The code is slightly refactored [2].

[1] https://github.com/dlang/dmd/blob/master/test/compilable/objc_interface_final_19654.d
[2] https://github.com/dlang/dmd/pull/11821/files#diff-e88d6d54d409ea665c7643a4a89e3fe01d8decf90042dbf8f11e905a48aabfecL1711

**Comment by jacob-carlborg** [comment]
Ping. Another month has passed.

**Review by thewilsonator — APPROVED** [review]
